osquery/tables/events/linux/containerd_events.cpp

```diff
@@ -0,0 +1,333 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/core/flags.h>
+#include <osquery/logger/logger.h>
+#include <osquery/registry/registry_factory.h>
+#include <osquery/sql/sql.h>
+
+#include <boost/algorithm/string/join.hpp>
+
+#include <osquery/events/eventpublisher.h>
+#include <osquery/events/eventsubscriber.h>
+#include <osquery/events/linux/containereventpublisher.h>
+
+namespace osquery {
+
+namespace {
+std::string labelsMapToString(
+    std::unordered_map<std::string, std::string> labels) {
+  std::string labels_string;
+  for (const auto& label_pair : labels) {
+    labels_string += label_pair.first + "=" + label_pair.second + ",";
+  }
+
+  if (!labels_string.empty()) {
+    labels_string.pop_back();
+  }
+  return labels_string;
+}
+} // namespace
+
+class ContainerdContainerEventSubscriber final
+    : public EventSubscriber<ContainerEventPublisher> {
+ public:
+  Status init() override;
+
+  Status Callback(const ECRef& ec, const SCRef& sc);
+};
+
+REGISTER(ContainerdContainerEventSubscriber,
+         "event_subscriber",
+         "containerd_container_events");
+
+Status ContainerdContainerEventSubscriber::init() {
+  auto subscription = createSubscriptionContext();
+  subscription->event_type_subscription =
+      BaseContainerEvent::Type::ContainerEvent;
+
+  subscribe(&ContainerdContainerEventSubscriber::Callback, subscription);
+  return Status::success();
+}
+
+Status ContainerdContainerEventSubscriber::Callback(const ECRef& ec,
+                                                    const SCRef& sc) {
+  if (ec->container_event->type_ != BaseContainerEvent::Type::ContainerEvent) {
+    return Status::failure(
+        "Received wrong event type: " +
+        std::to_string(static_cast<int>(ec->container_event->type_)));
+  }
+
+  Row r;
+  const auto& container_event =
+      dynamic_cast<const ContainerEvent&>(*ec->container_event);
+
+  r["container_id"] = container_event.container_id_;
+  r["timestamp"] = container_event.timestamp_;
+  r["namespace"] = container_event.namespace_name_;
+  r["topic"] = container_event.topic_;
+  r["image_name"] = container_event.image_name_;
+  r["runtime"] = container_event.runtime_.runtime_;
+  r["runtime_root"] = container_event.runtime_.runtime_root_;
+  r["criu_path"] = container_event.runtime_.criu_path_;
+  r["systemd_cgroup"] = INTEGER(container_event.runtime_.systemd_cgroup_);
+  r["snapshot_key"] = container_event.snapshot_key_;
+
+  std::string labels = labelsMapToString(container_event.labels_);
+
+  r["labels"] = labels;
+
+  add(r);
```

Comment:
do you know how noisy/firehouse are these events? does it make sense to use `addBatch(r)` instead? That way writes to rocksdb could be batched up

Reply:
Chiming in with delay: as far as I recall events come when there is a state change in the container (so container being created, started, stopped), so they don't happen often. At the same time, they should not be many.

And additional thing about `addBatch`: remember that one has to provide multiple rows in the call. An `addBatch` of a single row is the same of `add`.

Reply:
It seems like we can get multiple events in a single call to `ContainerEventPublisher::run` but they tend to be of different types and thus would require separate `addBatch` calls. If a large number if containers are started at precisely the same time it could make sense though.